### PR TITLE
Quickstep date support

### PR DESCRIPTION
**Pull request by rogersjeffreyl**
This PR adds TPCH syntax support for the interval keyword. Quickstep supports  the following syntax for interval  : 
interval  '1 year'.
This PR adds support for an alternative syntax
interval '1' year

### CODE REVIEW

**Comment by jianqiao** [comment]
It would be better to support all these keywords:
```
YEAR
MONTH
DAY
HOUR
MINUTE
SECOND
```

File: parser/SqlParser.ypp [hunk 1]
```diff
@@ -1648,6 +1649,16 @@ literal_value:
       YYERROR;
     }
   }
+  | TOKEN_INTERVAL TOKEN_STRING_SINGLE_QUOTED TOKEN_YEAR {
```
Comment (jianqiao) [inline parser/SqlParser.ypp]: Let's support all the datetime keywords. One way to do it:
```
| TOKEN_INTERVAL TOKEN_STRING_SINGLE_QUOTED datetime_unit {
  ...
}

datetime_unit:
  TOKEN_YEAR {
    $$ = new quickstep::ParseString(@1.first_line, @1.first_column, std::string("year"));
  }
  | TOKEN_MONTH {
    $$ = new quickstep::ParseString(@1.first_line, @1.first_column, std::string("month"));
  }
  ...
```
where `datetime_unit` is in the type group `%type <string_value_>`

File: parser/ParseLiteralValue.hpp [hunk 2]
```diff
@@ -289,8 +289,7 @@ class StringParseLiteralValue : public ParseLiteralValue {
    **/
   static bool ParseAmbiguousInterval(ParseString *value,
                                      StringParseLiteralValue **output);
-
```
Comment (zuyu) [inline parser/ParseLiteralValue.hpp]: I don't understand why removed this line?

File: query_optimizer/tests/execution_generator/Select.test [hunk 3]
```diff
@@ -656,7 +656,7 @@ SELECT SUM(EXTRACT(SECOND FROM value))
 FROM dates;
 --
 +-------------------------------+
-|SUM(EXTRACT(SECOND FROM value))|
+|SUM(EXTRACT(second FROM value))|
```
Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: I was wondering if there are some standards about whether we should capitalize all the units.

Comment (rogersjeffreyl) [inline query_optimizer/tests/execution_generator/Select.test]: @zuyu looked at https://docs.oracle.com/cd/B28359_01/server.111/b28286/intro002.htm
does not seem to indicate any about uppercase specification. However the examples given indicate capitalization

Comment (rogersjeffreyl) [inline query_optimizer/tests/execution_generator/Select.test]: @zuyu should we modify it to reflect upper case?

Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: Yes, I think the convention would be that all variables (i.e., table and attribute names) are lowercase, while all key words should be uppercase.

File: parser/ParseLiteralValue.hpp [hunk 4]
```diff
@@ -289,8 +289,7 @@ class StringParseLiteralValue : public ParseLiteralValue {
...
-  /**
+   /**
```
Comment (jianqiao) [inline parser/ParseLiteralValue.hpp]: Shift left one space here.

File: query_optimizer/tests/execution_generator/Select.test (hunk 3 above)
Comment (jianqiao) [inline query_optimizer/tests/execution_generator/Select.test]: Yes let's use upper case here.

File: parser/SqlParser.ypp [hunk 5]
```diff
@@ -1627,6 +1634,19 @@ literal_value:
       YYERROR;
     }
   }
+  | TOKEN_INTERVAL TOKEN_STRING_SINGLE_QUOTED datetime_unit {
+    quickstep::StringParseLiteralValue *parse_value;
+    const std::string& datetime_type_value = $3->value();
+    if (quickstep::StringParseLiteralValue::ParseAmbiguousInterval(&($2->append((" " +
+        datetime_type_value).c_str(), datetime_type_value.length() + 1)),
```
Comment (jianqiao) [inline parser/SqlParser.ypp]: Perhaps it's better to put this expression as the first argument to `ParseAmbiguousInterval` in one single line. I.e.
```
    if (quickstep::StringParseLiteralValue::ParseAmbiguousInterval(
        &($2->append((" " + datetime_type_value).c_str(), datetime_type_value.length() + 1)),
        &parse_value)) {
      $$ = parse_value;
    } else {
      ...
```

Also update `const std::string& datetime_type_value` to `const std::string &datetime_type_value`.

**Comment by jianqiao** [comment]
@rogersjeffreyl Hi Rogers: just put a few comments. Also please rebase the commits to resolve the conflicts then I can merge the PR.

**Comment by jianqiao** [comment]
LGTM. I'ill rebase the commits and then merge the PR.

**Pull request merged**